ddtrace/tracer/sampler.go

```diff
@@ -147,3 +156,245 @@ func (ps *prioritySampler) apply(spn *span) {
+	for _, v := range rules {
+		if v.err != nil {
+			log.Warn("ignoring rule %v: %v", v, v.err)
+		}
+		if v.Rate < 0.0 || v.Rate > 1.0 {
+			log.Warn("ignoring rule %v: rate is out of range", v)
+		}
+		validRules = append(validRules, v)
+	}
+	return validRules
+}
+
+func sampleRate() float64 {
+	rate := 1.0
+	v := os.Getenv("DD_TRACE_SAMPLE_RATE")
+	if v == "" {
+		return rate
+	}
+	r, err := strconv.ParseFloat(v, 64)
+	if err != nil {
+		log.Warn("using default rate %f because DD_TRACE_SAMPLE_RATE is invalid: %v", rate, err)
+		return rate
+	}
+	if r >= 0.0 && r <= 1.0 {
+		return r
+	}
+	log.Warn("using default rate %f because provided value is out of range: %f", rate, r)
+	return rate
+}
+
+func newRateLimiter(r float64) *rate.Limiter {
+	v := os.Getenv("DD_TRACE_RATE_LIMIT")
+	if v == "" {
+		return nil
+	}
+	l, err := strconv.ParseFloat(v, 64)
+	if err != nil {
+		log.Warn("using default rate limit because DD_TRACE_RATE_LIMIT is invalid: %v", err)
+		return nil
+	}
+	switch {
+	case l < 0.0:
+		return nil
+	case l == 0.0:
+		return rate.NewLimiter(0.0, 0)
+	case (r * l) < 1.0:
+		return rate.NewLimiter(rate.Limit(l), 1)
+	default:
+		return rate.NewLimiter(rate.Limit(l), int(math.Ceil(r*l)))
+	}
+}
+
+func (rs *rulesSampler) apply(span *span) bool {
+	matched := false
+	sr := 0.0
+	for _, v := range rs.rules {
+		if v.match(span) {
+			matched = true
+			sr = v.Rate
+			break
+		}
+	}
+	if !matched {
+		return false
+	}
+	// rate sample
+	span.SetTag("_dd.rule_psr", sr)
+	if !sampledByRate(span.TraceID, sr) {
+		span.SetTag(ext.SamplingPriority, ext.PriorityAutoReject)
+		return true
+	}
+	// global rate limit and effective rate calculations
+	defer rs.mu.Unlock()
+	rs.mu.Lock()
+	if ts := time.Now().Unix(); ts > rs.ts {
+		// update "previous rate" and reset
+		if ts-rs.ts == 1 && rs.total > 0 && rs.allowed > 0 {
+			rs.previousRate = float64(rs.allowed) / float64(rs.total)
+		} else {
+			rs.previousRate = 0.0
+		}
+		rs.ts = ts
+		rs.allowed = 0
+		rs.total = 0
+	}
+
+	rs.total++
+	// calculate effective rate, and tag the span
+	er := (rs.previousRate + (float64(rs.allowed) / float64(rs.total))) / 2.0
+	span.SetTag("_dd.limit_psr", er)
+	if !rs.limiter.Allow() {
+		span.SetTag(ext.SamplingPriority, ext.PriorityAutoReject)
+		return true
+	}
+	span.SetTag(ext.SamplingPriority, ext.PriorityAutoKeep)
+	rs.allowed++
+
+	return true
+}
+
+// SamplingRule is used for applying sampling rates to spans that match
+// the service name, operation or both.
+// It's recommended to use the helper functions (ServiceRule, OperationRule,
+// ServiceOperationRule) instead of directly creating a SamplingRule.
+type SamplingRule struct {
+	Service   *regexp.Regexp
+	Operation *regexp.Regexp
+	Rate      float64
+	err       error
+}
+
+// ServiceRule returns a *SamplingRule that applies the provided sampling rate
+// to spans that match the service name provided.
+// The value for service can include regular expression syntax.
```

Comment:
What we’ve discussed and agreed upon was that the helper functions would match the exact string. If anyone wants to use regular expressions they should use the struct. Otherwise it’s both confusing and takes us back to square one.

Helper funcs need to apply only for *exact* matching.

Reply:
https://github.com/DataDog/dd-trace-go/pull/539#issuecomment-559706267

Reply:
You can parse and escape regexp special characters with a simple algorithm.